src/crystal/system/wasm/wasi.cr

```diff
@@ -0,0 +1,54 @@
+require "./lib_wasi"
+
+module Crystal::System::Wasi
+  PREOPENS = [] of {String, String, LibWasi::Fd}
+
+  def self.init_preopens
+    # Skip stdin, stdout, and stderr, and count up until we reach an invalid file descriptor.
+    (3..).each do |fd|
+      stat = uninitialized LibWasi::Prestat
+      err = LibWasi.fd_prestat_get(fd, pointerof(stat))
+
+      break if err.badf?
+      unless err.success?
+        raise RuntimeError.from_os_error("fd_prestat_get", err)
+      end
+
+      next unless stat.tag.dir?
+
+      len = stat.value.dir.pr_name_len
+
+      name = String.new(len + 1) do |buffer|
+        err = LibWasi.fd_prestat_dir_name(fd, buffer, len)
+        raise RuntimeError.from_os_error("fd_prestat_dir_name", err) unless err.success?
+        buffer[len] = 0
+        len = LibC.strlen(buffer)
+        {len, 0}
+      end
+
+      path = ::Path[name].expand.to_s
+      PREOPENS << {path, path.ends_with?("/") ? path : path + "/", fd}
+    end
+
+    # Preopens added later take priority over preopens added earlier.
+    PREOPENS.reverse!
+    # Preopens of longer prefix take priority over shorter prefixes.
+    PREOPENS.sort_by! { |entry| -entry[0].size }
+  end
```

Comment:
would it make sense to write this as a `PREOPENS = begin ... end` expression instead? the separate class method and modifying the constant feels a bit like breaking a rule of some kind. (not a big deal though)

Reply:
Unfortunately this constant has to be initialized as early as possible, before any call to `LibC.open` is made. Constants are usually initialized on the first use and that might be too late. Is there a better way to ensure it is initialized without doing what I did?

Reply:
Wouldn't code of the form

```crystal
PRELOADS = begin
  preloads = {} of K=>V
  # init_preloads code
  preloads
end```

will run the block at the point where PRELOADS is declared and without having to modify the constant after it is initialized?

Even with lazy initialization, couldn't you force it to be initialized by referencing the constant in the next line?